src/mpfao/3D/tdympfao3D_ts.c

```diff
@@ -75,8 +77,15 @@ PetscErrorCode TDyMPFAOIFunction_Vertices_3DMesh(Vec Ul, Vec R, void *ctx) {
       cell_id_up = faces->cell_ids[fOffsetCell + 0];
       cell_id_dn = faces->cell_ids[fOffsetCell + 1];
 
+      den = 0.0;
+      if (cell_id_up>=0) den += tdy->rho[cell_id_up];
+      else               den += tdy->rho_BND[-cell_id_up-1];
+      if (cell_id_dn>=0) den += tdy->rho[cell_id_dn];
+      else               den += tdy->rho_BND[-cell_id_dn-1];
+      den *= 0.5;
+
       // Upwind the 'ukvr'
-      if (TtimesP[irow] < 0.0) { // up ---> dn
+      if (TtimesP[irow] + den*GravDis_ptr[face_id*num_subfaces + subface_id]< 0.0) { // up ---> dn
```

Comment:
Since we make use of `GravDis_ptr[face_id*num_subfaces + subface_id]` here and below, maybe we could use something like `g` or `gravity` or something that indicates that it's the appropriate value for the gravity vector.